src/uct/tcp/tcp_ep.c

```diff
@@ -788,6 +896,93 @@ ssize_t uct_tcp_ep_am_bcopy(uct_ep_h uct_ep, uint8_t am_id,
     return payload_length;
 }
 
+ucs_status_t uct_tcp_ep_am_zcopy(uct_ep_h uct_ep, uint8_t am_id, const void *header,
+                                 unsigned header_length, const uct_iov_t *iov,
+                                 size_t iovcnt, unsigned flags,
+                                 uct_completion_t *comp)
+{
+    uct_tcp_ep_t *ep       = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr  = NULL;
+    uct_tcp_ep_zcopy_ctx_t *ctx;
+    struct iovec *ctx_iov;
+    ucs_status_t status;
+
+    UCT_CHECK_IOV_SIZE(iovcnt, iface->max_iov, "uct_tcp_ep_am_zcopy");
+    UCT_CHECK_LENGTH(header_length, 0, iface->max_zcopy_hdr,
+                     "am_zcopy header");
+    UCT_CHECK_LENGTH(header_length + uct_iov_total_length(iov, iovcnt), 0,
+                     iface->rx_seg_size - sizeof(uct_tcp_am_hdr_t),
+                     "am_zcopy");
+
+    status = uct_tcp_ep_am_prepare(iface, ep, am_id, &hdr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_assertv(hdr != NULL, "ep=%p", ep);
+
+    ctx          = ucs_derived_of(hdr, uct_tcp_ep_zcopy_ctx_t);
+    ctx->iov_cnt = 0;
+    ctx_iov      = (struct iovec*)ctx->iov;
+
+    ctx_iov[ctx->iov_cnt].iov_base = hdr;
+    ctx_iov[ctx->iov_cnt].iov_len  = sizeof(*hdr);
+    ctx->iov_cnt++;
+
+    if (header_length != 0) {
+        ucs_assert(header != NULL);
+        ctx_iov[ctx->iov_cnt].iov_base = (void*)header;
+        ctx_iov[ctx->iov_cnt].iov_len  = header_length;
+        ctx->iov_cnt++;
+    }
+
+    ctx->iov_cnt  += uct_tcp_ep_iovec_fill_iov(&ctx_iov[ctx->iov_cnt], iov,
+                                              iovcnt, &ep->tx.length);
+    hdr->length    = ep->tx.length + header_length;
+    ep->tx.length  = hdr->length   + sizeof(*hdr);
+
+    ucs_assertv(ep->tx.length <= iface->rx_seg_size, "ep=%p", ep);
+
+    uct_iface_trace_am(&iface->super, UCT_AM_TRACE_TYPE_SEND, hdr->am_id,
+                       header, hdr->length, "SEND %p fd %d", ep, ep->fd);
+
+    iface->outstanding += ep->tx.length;
+
+    status = ucs_socket_sendv_nb(ep->fd, ctx_iov, ctx->iov_cnt,
```

Comment:
btw, would not be better to make `ucs_socket_sendv_nb` working with uct_iovs, and make all translation to iovs inside it ?